Angular: Deprecate the story component attribute

ThibaudAV opened the pull request:
Issue: https://github.com/storybookjs/storybook/issues/8673

TODO : 
- [x] Add MIGRATION.md info
- [x] change the current examples
- [x] check the doc

## What I did
- use same "angular preview render" in storyshots and in app/angular
- fix some exemples
- deprecate story component attribute
- update all exemples

## How to test

- Is this testable with Jest or Chromatic screenshots?
- Does this need a new example in the kitchen sink apps?
- Does this need an update to the documentation?

If your answer is yes to any of these, please make sure to include it in your PR.

<!--

Everybody: Please submit all PRs to the `next` branch unless they are specific to the current release. Storybook maintainers cherry-pick bug and documentation fixes into the `master` branch as part of the release process, so you shouldn't need to worry about this.

Maintainers: Please tag your pull request with at least one of the following:
`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

-->

ThibaudAV committed:
feat(angular): depreciate the story component attribute

ThibaudAV committed:
refactor(angular): take into account the `component` deprecation in angular docs

ThibaudAV committed:
refactor(angular): take into account the `component` deprecation in angular cli

ThibaudAV commented on app/angular/src/client/preview/angular-beta/RenderNgAppService.ts:
```diff
@@ -63,9 +66,27 @@ export class RenderNgAppService {
+        or
+        \`\`\`
+        export const Primary: Story = () => ({});
+        Primary.parameters = { component: AppComponent };
```
don't hesitate if you have a better suggestion :) 

There are other possibilities :

```
storiesOf(...)
  .addParameters({
    component: AppComponent,
  })
```
or : 
```
.add(
    'background component',
    () => ({
      props: {},
    }),
    {
      component: AppComponent,
    }
  )
```
🤷‍♂️

ThibaudAV commented:
@shilman I'm not sure what to put in MIGRATION.md 🤷‍♂️  Can I let you do it? 

some comments : 
- It's a mess in angular-cli examples ^^
- I did my best for the doc snippets. but I think it needs another look

ThibaudAV committed:
fix(angular): use @storybook/angular client rendering in storyshots

ThibaudAV committed:
fix(angular): declare only the compatible component

ignore class like doc-injectable example

ThibaudAV commented on addons/storyshots/storyshots-core/src/frameworks/angular/renderTree.ts:
```diff
@@ -7,35 +7,35 @@ import { BrowserDynamicTestingModule } from '@angular/platform-browser-dynamic/t
 import { NO_ERRORS_SCHEMA } from '@angular/core';
 import { addSerializer } from 'jest-specific-snapshot';
-import { initModuleData } from './helpers';
+import { RenderNgAppService } from '@storybook/angular/dist/client/preview/angular-beta/RenderNgAppService';
```
Avoids redefining the rendering like before.
if you have a (simple) idea to avoid the `/dist/` tell me :)

shilman committed:
MIGRATION: Deprecate Angular story component

shilman commented on app/angular/src/client/preview/angular-beta/RenderNgAppService.ts:
```diff
@@ -19,14 +21,72 @@ import { isComponentAlreadyDeclaredInModules } from './NgModulesAnalyzer';
+    } = storyFnAngular;
+
+    if (storyComponent) {
+      logger.warn(dedent`\`component\` story input is deprecated, and will be removed in Storybook 7.0.
```
Please use `util-deprecate`. That way it's easier to find all the instances before 7.0 and remove them.

Example: https://github.com/storybookjs/storybook/blob/master/lib/core/src/client/preview/loadCsf.ts#L9

ThibaudAV replied:
I thought, it was only for functions. It's done https://github.com/storybookjs/storybook/pull/13383/commits/6e6de1b357b28ef6cac13dc2ba7a09c731e8c11f#diff-57fd10a2b9f215006eeda934f6e927d27c8773e71e488b3c04d9a6168fe303c0R52

shilman committed:
Fix link

shilman commented on docs/snippets/angular/button-story-component-decorator.ts.mdx:
```diff
@@ -22,10 +22,6 @@ export default {
 
 const Template: Story<ButtonComponent> = (args: ButtonComponent) => ({
-  component: ButtonComponent,
-  moduleMetadata: {
```
Does this `moduleMetadata` also come for free now? If so, we also need to add that to MIGRATION. Let me know and I can do that.

ThibaudAV replied:
IMHO : no,  There is no change. It's just that it's redundant with the `component: ButtonComponent,` in `export default` 

In angular all components must be declared in only one module.
storybook/angular builds a module and declares `component` if necessary and also components listed in `moduleMetadata.declarations`.

shilman reviewed (changes requested):
This is so awesome @ThibaudAV 💯  Thanks for pushing it forward.

I wrote a MIGRATION and added a couple comments to the PR including one change request.

Would be great to get more @storybookjs/angular eyes on it to critique from an Angular standpoint. From a CSF standpoint it looks like a big improvement to me. I also appreciate the example cleanup.

One more thing, can we leave around an "old-style" story as a test case, if there's not one already? Or create one?

ThibaudAV committed:
refactor(angular): take into account the `component` deprecation in angular examples

ThibaudAV committed:
fix(angular): use utc to avoid time zone differences

shilman reviewed (approved):
Looking great. Awesome work @ThibaudAV 😍 

I'm hoping @kroeder and @Marklb can take a look before we merge.

shilman committed:
Merge branch 'next' into pr/13383

Marklb commented:
I haven't gone through all the snippet and example changes yet, but so far it looks good.

---
Not that is should be done in this PR, but I think this change increases the importance of getting the dynamically generated source implemented, like the React framework has. Now that the `StoryFnAngularReturnType` isn't as much of a self-contained representation of the Story, the source doesn't really tell the user anything about the Story.

For example, a Story that doesn't use a `template` and only relies on `addon-docs` for it's props, has nothing of importance in the Story tab. It doesn't even let the user know the name of the component class, at the least.
![image](https://user-images.githubusercontent.com/8348647/101737028-e7dec900-3a89-11eb-9885-1ff6ca1d18df.png)

shilman commented:
Thanks @Marklb 🙏 

To clarify a bit, there are two different "sources" here. One is `addon-storysource` which probably deserves to be deprecated in favor of `addon-docs` Source block and `addon-source` which was developed by Microsoft and visible on the "Source" tab here, and is OSS but currently unreleased AFAIK: https://fluidframework.com/playground/?path=/sourceCode/react-demos-draft-js--demo

gaetanmaisse commented on addons/storyshots/storyshots-core/package.json:
```diff
@@ -55,10 +55,12 @@
     "react-test-renderer": "^16.8.0 || ^17.0.0",
     "read-pkg-up": "^7.0.0",
     "regenerator-runtime": "^0.13.7",
+    "rxjs": "^6.6.3",
```
Maybe `rxjs` should be a peerDep too as non Angular users don't need this dep

gaetanmaisse commented on examples/angular-cli/src/stories/app.component.stories.ts:
```diff
@@ -8,10 +8,7 @@ export default {
+export const ComponentWithSeparateTemplate = () => ({});
 
-ComponentWithSeparateTemplate.storyName = 'Component with separate template';
+ComponentWithSeparateTemplate.storyName = 'Component with fullscreen parameters';
```
Intended rename?

ThibaudAV replied:
Yes previous name makes no sense IMHO

gaetanmaisse reviewed (changes requested):
It looks great, well done @ThibaudAV! Awesome to see some work on `@storybook/angular` :)

My only concern is about `@storybook/storyshots` because the changes made in this PR are maybe a breaking change 🤔 . We can see that all storyshots have been regenerated to match the outputs of the new "render" engine, which means users updating to 6.2 will need to do the same. 

I just raise this flag to discuss with you because I remember having introduced a breaking change in this addon in 5.x a few years ago (a youthful mistake🙈 ) and users were quite unhappy. I think @shilman also remembers 😁

shilman commented on app/angular/src/client/preview/angular-beta/RenderNgAppService.ts:
```diff
@@ -19,14 +21,75 @@ import { isComponentAlreadyDeclaredInModules } from './NgModulesAnalyzer';
+    } = storyFnAngular;
+
+    if (storyComponent) {
+      deprecate(
```
You need to declare a warning function at the top of the file like in the order code examples. Otherwise the deprecation warning prints every time you call it, rather than just on the first call

shilman commented on examples/angular-cli/src/stories/__snapshots__/addon-actions.stories.storyshot:
```diff
@@ -1,37 +1,47 @@
-  cfr={[Function CodegenComponentFactoryResolver]}
-  data={[Function Object]}
-  target={[Function ViewContainerRef_]}
+<storybook-wrapper
```
Good catch @gaetanmaisse on the snapshots. Is it easy to make this a non-breaking change?

shilman reviewed (changes requested):
Another pass 😂

ThibaudAV commented:
Thanks @gaetanmaisse you're right, but how do you do it?
It makes more sense to use the same `render` everywhere. Especially between test and preview no ? 
Can we warn the users to regenerate snapshots to this version ? 
Add something in MIGRATION ? 

@Marklb we can wait for you to finish review :)
And yes this is one of 2 subjects that I think to start next 
1. better support for Decorators https://github.com/storybookjs/storybook/issues/10272 or https://github.com/storybookjs/storybook/issues/12749
2. representation of the Story https://github.com/storybookjs/storybook/issues/10617

shilman commented:
@ThibaudAV I think we can add a breaking change if that's necessary. But I don't understand why it's needed.

ThibaudAV committed:
fixup! feat(angular): depreciate the story component attribute

ThibaudAV committed:
feat(angular): remove internal storybook properties in storyshot

Generates snapshots that will change less if the internal stroybook properties change

ThibaudAV commented:
@shilman 
IMHO : I think it would be better to change the snapshots. Otherwise there will be other problems

To generate the snapshot storyshot needs some information built by the "render".
Before there was a code duplication and so 2 versions of the render (one for addon/storyshot and one for stroybook/angular)
This is problematic because the stroyshot version does not benefit from the improvements and corrections of the main part (stroybook/angular), for some time now
It also makes tests that do not reflect exactly the story generated

If we continue with the same code as before the new internal storybook functionality will not work with storyshot anymore. and this will make the storyshot completely legacy. :/

To prevent snapshots from changing due to internal storybook modifications🤞
My last commit removes from snapshots some internal properties that may change in the future. 
And allows [snapshots to contain mainly the story code](https://github.com/storybookjs/storybook/pull/13383/files#diff-fc34dc3bbef46f5630b441c12de841d5182d0a69a704eb2e19c247a133f909e6L4-L21). 

What do you think about adding also a message in MIGRATION to inform that storyshot is improving to contain less internal sotrybook properties. And that it will be necessary to regenerate them 🙈  but it's for a good reason

I have a little talk with @gaetanmaisse  who will surely be able to complete some information

I don't know if it's reassuring or not but when angular or angular-material changes minor version and you have test snapshot on your project it's not uncommon to have to regenerate them

In any case, I thank you for accompanying me in these changes, it is very nice to have your opinion. 😌

gaetanmaisse commented:
I think the final result we have for story snapshots is a huge improvement because:
 - it now really matches what is rendered when running Storybook
 - all Storybook renderer specific properties aren't included in the snapshots anymore, so we can now update the renderer internal plumbing without causing any update of the snapshots 👌🏻 

So the remaining question is: are we ok to ask SB users to update all their snapshots when updating to 6.2? Or must we wait for 7.0 to do any change in `@storybook/addon-storyshot`?

I would prefer to merge and release this in 6.2 as it's part of major improvements of SB for Angular 🤷🏻

gaetanmaisse reviewed (approved):
The code LGTM, if we agree on making this land in 6.2 we have to add a section about snapshots update in MIGRATION.md

shilman committed:
Document Angular snapshot migration

shilman committed:
Merge branch 'next' into pr/13383

shilman commented:
@ThibaudAV @gaetanmaisse thanks for getting this over the line. I've updated the MIGRATION instructions as I understand them:

> We've updated the Angular storyshots format in 6.2, which is technically a breaking change. Apologies to semver purists: if you're using storyshots, you'll need to [update your snapshots](https://jestjs.io/docs/en/snapshot-testing#updating-snapshots). The new format hides the implementation details of `@storybook/angular` so that we can evolve its renderer without breaking your snapshots in the future.

I'm OK with this compromise, but since this is a breaking change in a minor release I'm going to give the rest of the maintainers a day or two to chime in before merging.

manuelmeister commented on MIGRATION.md:
```diff
@@ -143,9 +146,42 @@
+
+#### New Angular renderer
+
+We've rewritten the Angular renderer in Storybook 6.2. It's meant to be entirely backwards compatible, but if you need to use the legacy renderer it's still available via a [parameter](https://storybook.js.org/docs/react/writing-stories/parameters). To opt out of the new renderer, add the following to `.storybook/preview.ts`:
```
Why is this link to react? Wouldn't it make sense to link to angular?

shilman committed:
Update MIGRATION.md

Co-authored-by: Manuel Meister <news.manuelsworld@gmail.com>

ThibaudAV committed:
Merge branch 'next' into angular-parameters-component

# Conflicts:
#	MIGRATION.md

shilman committed:
Merge branch 'next' into pr/13383

shilman committed:
Fix angular storyshots

ThibaudAV commented on app/angular/src/client/index.ts:
```diff
@@ -9,6 +9,8 @@ export {
   raw,
 } from './preview';
 
+export { RenderNgAppService } from './preview/angular-beta/RenderNgAppService';
```
I hadn't dared to export this service in the public index. ^^

shilman replied:
I'm not sure this is the right way. However, we recently introduced ESM into the codebase and it broke things: #13013. If there's a less aggressive way to do it, I'm also open to that!

shilman committed:
Merge branch 'next' into pr/13383

shilman committed:
Angular: Special `renderer` entry for storyshots

shilman commented on addons/storyshots/storyshots-core/src/frameworks/angular/renderTree.ts:
```diff
@@ -5,7 +5,7 @@ import { TestBed } from '@angular/core/testing';
 import { BrowserDynamicTestingModule } from '@angular/platform-browser-dynamic/testing';
 import { addSerializer } from 'jest-specific-snapshot';
-import { RenderNgAppService } from '@storybook/angular';
+import { RenderNgAppService } from '@storybook/angular/renderer';
```
@ThibaudAV trying this as something less aggressive 😄

ThibaudAV replied:
yes but I can see that it doesn't seem simple 💪 . Tell me if you need help with this PR

shilman committed:
Angular: Fix renderer export

shilman committed:
Update renderer.js

shilman committed:
Update renderer.js

shilman commented:
@ThibaudAV I think this is ready to merge. WDYT?

ThibaudAV commented:
@shilman 🤞

shilman commented:
cc @mandarini
